### ISSUE TITLE
mypy types set of `StrEnum` as set of `str`

### ISSUE THREAD

**Issue by injust** [python/mypy#18380]
**Bug Report**

mypy types a set of `StrEnum` as `set[str]`. 

**To Reproduce**

```python
from enum import Enum, StrEnum

class Tag(Enum):
    pass

class StrTag(StrEnum):
    pass

reveal_type(set(Tag))  # note: Revealed type is "builtins.set[__main__.Tag]"
reveal_type(set(StrTag))  # note: Revealed type is "builtins.set[builtins.str]"
```

https://mypy-play.net/?mypy=latest&python=3.13&gist=237de7c0610830acd5a5104d18541e3e

**Expected Behavior**

mypy should type `set(StrTag)` as `set[StrTag]`, not `set[str]`.

**Actual Behavior**

```
main.py:10: note: Revealed type is "builtins.set[builtins.str]"
```

**Your Environment**

- Mypy version used: 1.14.0
- Python version used: 3.13

**Comment by sobolevn** [comment]
This happens because `str` is also `Iterable` and `str` is closer in the `MRO` than `Enum`. Adding `def __iter__(self) -> Self: ...` to `StrEnum` solves the issue. But, I am not sure that this is the correct / full fix.

**Comment by injust** [comment]
I think https://github.com/python/cpython/issues/128659 is the root cause.

**Comment by sobolevn** [comment]
I don't think so: mypy does not use runtime analysis. So, these are two different cases.

**Comment by hauntsaninja** [comment]
This was fixed by #18587
